### PR TITLE
feat: Draft Add sections framework

### PR DESCRIPTION
**Pull request by talledodiego**
This PR contains a first draft on sections level.
The structure given considers the section (for now only the generic section has been implemented) object that contains a combination of geometries (e.g. surfaces, points) and materials (constitutive relation). The section calculator is the object implementing the different computation algorithms. The integrator object is the one necessary for doing the integration of stresses in the section: two integrators have been implemented, Marin (default) and Fibre integration. The results are stored as proper dataclasses.
The implementation relys on shapely package.

Some limited usecase examples are reported.

1. Importing the necessary functions

```python
from structuralcodes.materials.constitutive_laws import ElasticPlastic
from structuralcodes.materials.concrete import ConcreteMC2010
from structuralcodes.materials.reinforcement import ReinforcementMC2010
from structuralcodes.sections._generic import GenericSection
from structuralcodes.sections._geometry import add_reinforcement, add_reinforcement_line, SurfaceGeometry
from shapely import Polygon
```

2. Creating material objects (note: the section framework supports both constitutive laws or material classes from where the constitutive law is available)

```python
concrete = ConcreteMC2010(30)
reinforcement = ReinforcementMC2010(410, 200000, 410, 0.0675)
```

3. Create a rectangular section with some reinforcement

```python
rect = Polygon(((-100,-250),(100,-250),(100,250),(-100,250)))
geo = SurfaceGeometry(rect, concrete)
geo = add_reinforcement(geo, (-60, 210), 16, reinforcement)
geo = add_reinforcement(geo, (60, 210), 16, reinforcement)
```
The output is the following:
image

4. Different ways of adding reinforcement:
Specifying a line and a number of reinforcements (spacing is automatically computed):

```python
add_reinforcement_line(geo, (-60,-210), (60,-210), 16, reinforcement, n = 4)
```
image
Specifying a line and the spacing of bars (number is automatically computed):

```python
add_reinforcement_line(geo, (-60,-210), (60,-210), 16, reinforcement, s = 30)
```
image
The function can be also nested

```python
geo = add_reinforcement_line(
    add_reinforcement_line(geo, (-60, -210), (60, -210), 16, reinforcement, n = 2),
    (-60,-210),
    (60,-210),
    8,
    reinforcement,
    s = 20,
    first = False,
    last = False
)
```
image
Also, both spacing and number can be given:

```python
add_reinforcement_line(
    add_reinforcement_line(geo, (60, -210), (60, 210), 12, reinforcement, n = 2, s = 100), 
    (-60, -210), 
    (-60, 210), 
    12, 
    reinforcement, 
    n = 2, 
    s = 100
)
```
image
5. Translate and other editing are included

```python
geo = geo.translate(-100,-100)
```
6. Create the section (if integrator is not specified, Marin integration is used)

```python
sec = GenericSection(geo, 'mySection', integrator = 'Fiber')
```
7. Compute gross properties (e.g. Area)

```python
print(sec.gross_properties.area)
```
8. Compute bending strength
```python
res = sec.section_analyzer.calculate_bending_strength(theta = 0, n = 0)
print(res.m_x)
```
9. Compute moment Curvature
```python
res = sec.section_analyzer.calculate_moment_curvature(theta=0, n=0)
import matplotlib.pyplot as plt
plt.plot(res.chi, -res.my)
```
10. Analyze a steel section (like I-shaped)
```python
fyk = 355
Es = 206000
Eh = 0.0
eps_su = 10.0e-2 #plastic strength
eps_su = fyk / Es #elastic strength

# H section
h = 300
b = 150
tf = 10.7
tw = 7.1
steel = ElasticPlastic(E=Es, fy=fyk, Eh=Eh, eps_su=eps_su)
# Geometry
sec_geom = Polygon(
    (
        (-b/2, -h/2),
        (b/2, -h/2),
        (b/2, -h/2+tf),
        (tw/2, -h/2+tf),
        (tw/2, h/2 - tf),
        (b/2, h/2-tf),
        (b/2, h/2),
        (-b/2, h/2),
        (-b/2, h/2 - tf),
        (-tw/2, h/2 - tf),
        (-tw/2, -h/2 + tf),
        (-b/2, -h/2 + tf),
    )
)
# Trick: create a CompundGeometry with just one SurfaceGeometry to make algorithms work
geo = CompoundGeometry([SurfaceGeometry(sec_geom, steel)])
# Section
sec = GenericSection(geo)

res_fiber = sec.section_analyzer.calculate_bending_strength(theta=0, n=0)
print(f'Mx = {res_fiber.m_x * 1e-6:.2f} kNm')
# Expected bending strength
J = 2/12 * (b * tf**3) + 2 * (tf*b) * (h/2-tf/2)**2 + 1/12 * (tw * (h-2*tf)**3)
W = J/(h/2)
print(f'Expected Mx,el = {fyk * W * 1e-6:.2f} kNm')
Wpl = 2 * (tf * b * (h/2 - tf/2) + (h/2-tf) * tw * (h/2-tf)/2)
print(f'Expected Mx,pl = {fyk * Wpl * 1e-6:.2f} kNm')
```

### CODE REVIEW

File: structuralcodes/core/base.py [hunk 1]
```diff
@@ -48,3 +51,114 @@ def name(self):
...
+    def get_stress(self, eps: float) -> float:
+        """Each constitutive law should provide a method to return the
+        stress given the strain level"""
+
+    @abc.abstractmethod
+    def get_tangent(self, eps: float) -> float:
+        """Each constitutive law should provide a method to return the
+        tangent at a given strain level"""
+
+    def get_secant(self, eps: float) -> float:
```
Comment (mortenengen) [inline structuralcodes/core/base.py:85]: Should `get_secant` call `get_tanget` for `eps = 0.0` to avoid a `ZeroDivisionError`?

File: structuralcodes/core/base.py [hunk 2]
```diff
@@ -48,3 +51,114 @@ def name(self):
...
+    """Abstract base class for a cross secion.
+    The section is defined by local axes y and z (mapped to x and y cartesian
+    plane respectively)
+    """
+
+    section_counter: t.ClassVar[int] = 0
+
+    def __init__(self, name: t.Optional[str] = None) -> None:
+        self.id = self.section_counter
+        self._name = name if name is not None else "Section_{}".format(self.id)
```
Comment (mortenengen) [inline structuralcodes/core/base.py:102]: Consider `f'Section_{self.id}'` instead of the old `.format` way 😃

File: structuralcodes/sections/_section_results.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    i_yy: float = 0  # strong axis
+    i_zz: float = 0  # weak axis
+    i_yz: float = 0
+    # TODO: principal axes i_11 i_22 and angle
+
+    # section flexural rigidity
+    ei_xx: float = 0
+    ei_yy: float = 0
+
+    def __repr__(self, fmt: str = ".3f") -> str:
```
Comment (mortenengen) [inline structuralcodes/sections/_section_results.py:42]: This should be the `__str__` method since it prints an informal string representation of the object. I am not sure how common it is to have either `__str__` or `__repr__` with additional arguments though...

File: structuralcodes/materials/concrete/_concrete.py [hunk 4]
```diff
@@ -43,3 +48,19 @@ def fck(self, fck: float) -> None:
...
         """Each concrete should define its own _reset_attributes method
         This is because fck setting, reset the object arguments"""
+
+    @property
+    def stress_strain(self) -> ConstitutiveLaw:
+        """Returns the constitutive law object"""
+        return self._stress_strain
+
+    @stress_strain.setter
+    def stressStrain(self, stress_strain: ConstitutiveLaw) -> None:
```
Comment (mortenengen) [inline structuralcodes/materials/concrete/_concrete.py:58]: Please use snake_case for the method name 🐍

File: structuralcodes/materials/constitutive_laws.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    __materials__: t.Tuple[str] = (
+        'steel',
+        'rebars',
+    )
+
+    def __init__(
+        self,
+        E: float,
+        fy: float,
+        Eh: t.Optional[float] = None,
```
Comment (mortenengen) [inline structuralcodes/materials/constitutive_laws.py:45]: We could avoid the `if` on L54-57 by saying:
```python
...
Eh: float = 0.0,
...
```

File: structuralcodes/materials/constitutive_laws.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        name = name if name is not None else "ElasticPlasticLaw"
+        super().__init__(name=name)
+        self._E = E
+        self._fy = fy
+        if Eh is not None:
+            self._Eh = Eh
+        else:
+            self._Eh = 0.0
+        self._eps_su = eps_su
+        self._eps_sy = fy / E
```
Comment (mortenengen) [inline structuralcodes/materials/constitutive_laws.py:59]: Consider catching the `ZeroDivisionError` in case `E` is accidentally set to zero.

File: structuralcodes/materials/constitutive_laws.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        delta_sig = self._fy * (1 - self._Eh / self._E)
+        sig[sig < -self._fy] = eps[sig < -self._fy] * self._Eh - delta_sig
+        sig[sig > self._fy] = eps[sig > self._fy] * self._Eh + delta_sig
+        if self._eps_su is not None:
+            sig[eps > self._eps_su] = 0
+            sig[eps < -self._eps_su] = 0  # pylint: disable=E1130
+        return sig
+
+    def get_tangent(self, eps: ArrayLike) -> ArrayLike:
+        """Return the tanet for given strain"""
```
Comment (mortenengen) [inline structuralcodes/materials/constitutive_laws.py:74]: Typo tanet -> tangent

File: structuralcodes/materials/constitutive_laws.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        self._E = E
+        self._fy = fy
+        if Eh is not None:
+            self._Eh = Eh
+        else:
+            self._Eh = 0.0
+        self._eps_su = eps_su
+        self._eps_sy = fy / E
+
+    def get_stress(self, eps: ArrayLike) -> ArrayLike:
```
Comment (mortenengen) [inline structuralcodes/materials/constitutive_laws.py:61]: Nice 👍

File: structuralcodes/materials/constitutive_laws.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+            self._y = np.concatenate((-np.flip(y), y))
+        else:
+            # User gave both positive and negative parts
+            self._x = x
+            self._y = y
+
+    def get_stress(self, eps: ArrayLike) -> ArrayLike:
+        """Return the stress given strain"""
+        eps = np.atleast_1d(np.asarray(eps))
+        sig = np.interp(eps, self._x, self._y, left=0, right=0)
```
Comment (mortenengen) [inline structuralcodes/materials/constitutive_laws.py:159]: Here, `np.interp` returns `0` after the last values to the left and right, and not a constant value as described in the docstring. Perhaps to stick to contant value and slightly rewrite the call to `np.interp`?

File: test_generic.ipynb
Comment (mortenengen) [inline test_generic.ipynb:1]: Remember to remove this file before we merge.

File: test_rectangular.ipynb
Comment (mortenengen) [inline test_rectangular.ipynb:1]: Remember to remove this file before we merge.

File: tests/test_constitutive_laws.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+from structuralcodes.materials.constitutive_laws import Elastic
+from structuralcodes.materials.constitutive_laws import ElasticPlastic
+from structuralcodes.materials.constitutive_laws import ParabolaRectangle
+from structuralcodes.materials.constitutive_laws import UserDefined
+
+
+@pytest.mark.parametrize(
+    'E, strain, expected',
+    [(210000, 0.003, 210000 * 0.003), (200000, 0.002, 200000 * 0.002)],
```
Comment (mortenengen) [inline tests/test_constitutive_laws.py:15]: Consider to add test of negative value.

File: structuralcodes/sections/_generic.py [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+# to think: dataclass or class?
+# Note that some things are already computed (like area) by shapely
+# like: polygon.area, polygon.centroid, etc.
+
+# For now dataclass, if we need convert to regular class,
+# init commented for now
+
+
+class Geometry:
+    '''Base class for a geometry object'''
```
Comment (mortenengen) [inline structuralcodes/sections/_generic.py:32]: Not crucial, but in other places we tend to use double quotes for the docstrings and single ones for inline strings.

File: structuralcodes/sections/_generic.py [hunk 12]
```diff
@@ -0,0 +1,461 @@
...
+    if abs(v[0]) > 1e-8:
+        # it is a non vertical line
+        tg = v[1] / v[0]
+        x1 = bbox[0] - 1e-3
+        x2 = bbox[2] + 1e-3
+        y1 = point[1] + (x1 - point[0]) * tg
+        y2 = point[1] + (x2 - point[0]) * tg
+    else:
+        # it is a near-vertical line
+        # atg is almost zero
```
Comment (mortenengen) [inline structuralcodes/sections/_generic.py:154]: Minor one: should we call it ctg or similar since it refers to cotangent?

File: structuralcodes/sections/_generic.py [hunk 13]
```diff
@@ -0,0 +1,461 @@
...
+    '''Creates a line from point and angle within the bounding
+    box'''
+    # create a unit vector defining the line
+    v = (np.cos(theta), np.sin(theta))
+
+    # check if the line is vertical to avoid div by zero
+    if abs(v[0]) > 1e-8:
+        # it is a non vertical line
+        tg = v[1] / v[0]
+        x1 = bbox[0] - 1e-3
```
Comment (mortenengen) [inline structuralcodes/sections/_generic.py:148]: Why are we subtracting and adding the small number to the bounds of the bounding box? I just don't get it 😃

File: structuralcodes/sections/_generic.py [hunk 14]
```diff
@@ -0,0 +1,461 @@
...
+        if line.intersects(self.polygon):
+            result = split(self.polygon, line)
+            # divide polygons "above" and "below" line
+            for geom in result.geoms:
+                if LinearRing((line.coords[0],
+                               line.coords[1],
+                               geom.centroid.coords[0])).is_ccw:
+                    above_polygons.append(geom)
+                else:
+                    below_polygons.append(geom)
```
Comment (mortenengen) [inline structuralcodes/sections/_generic.py:248]: Nice check for above and below 👍

File: structuralcodes/sections/_generic.py [hunk 15]
```diff
@@ -0,0 +1,461 @@
...
+class SurfaceGeometry:
+    '''Class for a surface geometry with material.
+
+    Basically it is a wrapper for shapely polygon
+    including the material (and other parameters needed)
+    As a shapely polygon it can contain one or more holes'''
+
+    def __init__(
+        self,
+        poly: Polygon,
```
Comment (mortenengen) [inline structuralcodes/sections/_generic.py:173]: Instead of having a `from_points()` method, we could perhaps have

```python
...
poly: t.Union[Polygon, t.Tuple[Point], t.Tuple[t.Tuple[float]]],
...
```

to be able to initialize the `SurfaceGeometry` in a similar manner as the `Polygon` class?

File: structuralcodes/sections/_generic.py [hunk 16]
```diff
@@ -0,0 +1,461 @@
...
+        ii. PointGeometry(Geometry) that is a point with a mat
+        iii. SurfaceGeometry(Geometry) that is a polygon with a mat
+        iv. CompoundGeometry(Geometry) that is a set of geometries
+
+    For compound we could implement simple + method so that geo1 + geo2
+    returns a compunt.
+    So this method could become simply:
+
+    bar = Polygon(xxxx)
+    return geo + bar
```
Comment (mortenengen) [inline structuralcodes/sections/_generic.py:373]: Implementing the [`__add__`](https://docs.python.org/3/reference/datamodel.html#emulating-numeric-types) method so that we can add reinforcement by simply `geom + bar` would be a nice feature indeed 👍

File: structuralcodes/sections/_generic.py [hunk 17]
```diff
@@ -0,0 +1,461 @@
...
+
+    def __init__(
+        self,
+        geometry: SurfaceGeometry | CompoundGeometry,
+        name: t.Optional[str] = None
+    ) -> None:
+        if name is None:
+            name = 'GenericSection'
+        super().__init__(name)
+        self.section_analyzer: GenericSectionCalculator(self)
```
Comment (mortenengen) [inline structuralcodes/sections/_generic.py:389]: Here we should also set the attribute, not only annotate the type of it 👍

File: structuralcodes/sections/_generic.py [hunk 18]
```diff
@@ -0,0 +1,461 @@
...
+    bar = Polygon(xxxx)
+    return geo + bar
+    """
+    raise NotImplementedError
+
+
+class GenericSection(Section):
+    """This is the implementation of the generic class section"""
+
+    def __init__(
```
Comment (mortenengen) [inline structuralcodes/sections/_generic.py:381]: Could there be a situation where the user might want to provide a `SectionCalculator` when initializing the `Section`? E.g. providing a custom calculator.

**Comment by talledodiego** [comment]
I've commented all your comments. 
Almost everything is done, except some minor things that will be done later.

File: structuralcodes/materials/constitutive_laws.py (hunk 9 above)
Comment (talledodiego) [reply structuralcodes/materials/constitutive_laws.py:159]: Discussed with @mortenengen: we would add the flag with following options:

1. drop to zero
2. constant
3. tangent
4. secant

File: tests/test_constitutive_laws.py (hunk 10 above)
Comment (mortenengen) [reply tests/test_constitutive_laws.py:15]: Actually, @talledodiego, I was thinking of testing the response to negative strains 😃

**Comment by mortenengen** [comment]
- [ ] Drop `functools.cache` for version 3.8

File: structuralcodes/sections/_section_results.py (hunk 3 above)
Comment (mortenengen) [reply structuralcodes/sections/_section_results.py]: I like the `__format__` way of approaching this :smiley:

File: structuralcodes/sections/_section_integrators.py [hunk 19]
```diff
@@ -0,0 +1,442 @@
...
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+from math import atan2, cos, sin
+
+from shapely.geometry.polygon import orient
+from shapely import MultiLineString, Polygon, MultiPolygon
+
+import triangle
```
Comment (mortenengen) [inline structuralcodes/sections/_section_integrators.py:13]: Remember to add this dependency to pyproject.toml 👍

File: structuralcodes/sections/_section_integrators.py [hunk 20]
```diff
@@ -0,0 +1,442 @@
...
+        angle = atan2(strain[2], strain[1])
+
+        rotated_geom = geo.rotate(angle)
+        # 2. Get y coordinate of neutral axis in this new CRS
+        strain_rotated = [strain[0], (strain[2] ** 2 + strain[1] ** 2) ** 0.5]
+
+        # 3. For each SurfaceGeometry on the CompoundGeometry:
+        for g in rotated_geom.geometries:
+            # 3a. get coefficients and strain limits from constitutive law
+            strains, coeffs = g.material.__marin__(strain_rotated)
```
Comment (mortenengen) [inline structuralcodes/sections/_section_integrators.py:102]: Here we should catch the `AttributeError` that is raised if the `__marin__` method is not defined.

File: structuralcodes/sections/_section_integrators.py [hunk 21]
```diff
@@ -0,0 +1,442 @@
...
+                if y0 > y1:
+                    y0, y1 = y1, y0
+                bbox = g.polygon.bounds
+                line0 = create_line_point_angle((0, y0), 0, bbox)
+                line1 = create_line_point_angle((0, y1), 0, bbox)
+                lines = MultiLineString((line0, line1))
+                # intersection
+                result = g.split_two_lines(lines=lines)
+
+                def get_input_polygon(polygon, coeffs):
```
Comment (mortenengen) [inline structuralcodes/sections/_section_integrators.py:118]: Why is this function defined within the scope of the method?

File: structuralcodes/materials/constitutive_laws.py [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+        self._eps_u = -abs(eps_u)
+        self._n = n
+
+    def get_stress(self, eps: ArrayLike) -> ArrayLike:
+        """Return the stress given strain."""
+        eps = np.atleast_1d(np.asarray(eps))
+        # Parabolic branch
+        sig = self._fc * (1 - (1 - (eps / self._eps_0)) ** self._n)
+        # Rectangle branch
+        sig[eps < self._eps_0] = self._fc
```
Comment (sshaave) [inline structuralcodes/materials/constitutive_laws.py:112]: Hi. Im a bit late to the party, but shouldn't the operator be flipped such that values where eps > self._eps_0 is set equal to self._fc?

File: structuralcodes/sections/_section_results.py (hunk 3 above)
Comment (talledodiego) [reply structuralcodes/sections/_section_results.py]: Included the `__format__` way of approaching it.

File: structuralcodes/materials/constitutive_laws.py (hunk 9 above)
Comment (talledodiego) [reply structuralcodes/materials/constitutive_laws.py]: Implemented this feature.
It is still needed to discuss one topic: the ultimate strain (the strain that defines a ultimate strain condition for sectional analysis) should be defined by user? Or anyhow taken as the last value of x given as input?

File: tests/test_constitutive_laws.py (hunk 10 above)
Comment (talledodiego) [reply tests/test_constitutive_laws.py:15]: done.

**Review by talledodiego — COMMENTED** [review]
Reviewd further comments.
In commit I am submitting those issues are solved, together with some bug fixes that came after some intensive testing.

File: structuralcodes/materials/constitutive_laws.py (hunk 22 above)
Comment (mortenengen) [reply structuralcodes/materials/constitutive_laws.py:117]: `eps_0` is negative, and forced negative in `__init__`, but we should also make sure that the user gets a heads up about this behaviour. Please add a docstring for the class :smiley:

**Comment by talledodiego** [comment]
@mortenengen in 3.8 tests fail because of cache not being found in module functool. Can you please assist me with this?

**Comment by mortenengen** [comment]
> @mortenengen in 3.8 tests fail because of cache not being found in module functool. Can you please assist me with this?

Fixed in 6d8521d 😃

**Comment by talledodiego** [comment]
@mortenengen with [9a9d2c7](https://github.com/fib-international/structuralcodes/pull/27/commits/9a9d2c7988b472a41b5eee9b25e706cc83e03530) I included several tests on steel structures with I section. I have a problem: with parametrize if I un-comment more than 2-3 lines for possible testing the make test freezes.. I don't know if this a memory limit for my PC. Any suggestion?

**Comment by mortenengen** [comment]
> @mortenengen with [9a9d2c7](https://github.com/fib-international/structuralcodes/pull/27/commits/9a9d2c7988b472a41b5eee9b25e706cc83e03530) I included several tests on steel structures with I section. I have a problem: with parametrize if I un-comment more than 2-3 lines for possible testing the make test freezes.. I don't know if this a memory limit for my PC. Any suggestion?

This was a strange one. There were tests that were either failing or freezing, and it was due to polygons that were joined through unions, but were still disconnected. Taking the union of disconnected geometries produces `MultiPolygon` and not `Polygon`, and therefore, the constructor of `SurfaceGeometry` raised a `TypeError`. I used the `set_precision` function from `shapely` to set the precision of the geometries to a very low number, and this made the tests pass 😃

**Comment by mortenengen** [comment]
In 1d64e13 I added constitutive laws to the `Reinforcement` class and swapped the `ElasticPlastic` objects with `ReinforcementMC2010` in relevant tests.
